File: api/controllers/usersController.js
const bcrypt = require("bcryptjs");
const asyncHandler = require("express-async-handler");
const CustomError = require("../utils/customError");
const { handleValidationErrors } = require("../utils/validator");
const { PrismaClient } = require("@prisma/client");
const prisma = new PrismaClient();

const getUserInfo = asyncHandler(async (req, res) => {
  const user = await prisma.user.findUnique({
    where: { id: req.user.id },
    select: {
      id: true,
      username: true,
      posts: {
        select: { published: true },
      },
      comments: {
        select: { id: true },
      },
    },
  });

  if (!user) {
    throw new CustomError(404, "User could not be found or does not exist.");
  }

  const publishedPostsCount = user.posts.filter((post) => post.published).length;
  const draftsCount = user.posts.length - publishedPostsCount;
  const commentsCount = user.comments.length;

  const userInfo = {
    id: user.id,
    username: user.username,
    isAuthor: publishedPostsCount > 0,
    posts: publishedPostsCount,
    drafts: draftsCount,
    comments: commentsCount,
  };

  return res.status(200).json(userInfo);
});

const createUser = asyncHandler(async (req, res) => {
  handleValidationErrors(req);

  const hashedPassword = await bcrypt.hash(req.body.password, 10);

  await prisma.user.create({
    data: {
      username: req.body.username,
      password: hashedPassword,
      isAuthor: Boolean(req.body.isAuthor).valueOf(), //ensures value is boolean
    },
  });

  res.status(201).json({ message: "User successfully created." });
});

const deleteUser = asyncHandler(async (req, res) => {
  const { password } = req.body;
  if (!req.user || !password) {
    return res.status(400).json({ message: "Password is required." });
  }

  const user = await prisma.user.findUnique({ where: { id: req.user.id } });
  if (!user) {
    console.error("User not found, ID: ", req.user.username);
    throw new CustomError(404, "User not found.");
  }

  const isPasswordValid = await bcrypt.compare(password, user.password);
  if (!isPasswordValid) {
    console.error("User deletion failed: Incorrect password for user:", req.user.username);
    return res.status(404).json({ message: "Incorrect password." });
  }

  await prisma.user.delete({
    where: { id: req.user.id },
  });

  res.status(200).json({ message: "User successfully deleted." });
});

module.exports = {
  getUserInfo,
  createUser,
  deleteUser,
};

File: server_flask/app/controllers/users_controller.py
from flask import jsonify, g
from app.models import User
from app import db
import bcrypt
from app.utils.auth_utils import generate_token, jwt_user_required


def create_user(data):
    username = data.get("username")
    password = data.get("password")
    is_author = data.get("isAuthor", False)

    hashed_pw = bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode(
        "utf-8"
    )
    user = User(username=username, password=hashed_pw, is_author=is_author)
    db.session.add(user)
    db.session.commit()

    return jsonify({"message": "User created!"}), 201


def login(data):
    username = data.get("username")
    password = data.get("password")

    user = User.query.filter_by(username=username).first()
    if user and bcrypt.checkpw(password.encode("utf-8"), user.password.encode("utf-8")):
        token = generate_token(user)
        return jsonify({"access_token": token}), 200

    return jsonify({"message": "Invalid credentials"}), 401


@jwt_user_required
def get_user_info():
    user = g.current_user

    published_posts_count = len([post for post in user.posts if post.published])
    drafts_count = len(user.posts) - published_posts_count
    comments_count = len(user.comments)
    is_author = published_posts_count > 0

    user_data = {
        "id": user.id,
        "username": user.username,
        "isAuthor": is_author,
        "posts": published_posts_count,
        "drafts": drafts_count,
        "comments": comments_count,
    }

    return jsonify(user_data), 200


def delete_user():
    user = g.current_user
    db.session.delete(user)
    db.session.commit()
    return jsonify({"message": "User deleted"}), 200
